server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -531,30 +552,28 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
 
         @Override
         public void onResponse(Releasable releasable) {
-            try {
-                assert replica.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
-                final ReplicaResult replicaResult = shardOperationOnReplica(replicaRequest.getRequest(), replica);
+            assert replica.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
+            shardOperationOnReplica(replicaRequest.getRequest(), replica, ActionListener.wrap((replicaResult) ->
```

Comment:
what if `shardOperationOnReplica` throws an exception? We're not releasing the releasable then?